src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,188 @@
+            if (currentNode.next != null) {
+                currentNode.next.prev = newNode;
+            }
+            currentNode.next = newNode;
+            newNode.prev = currentNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T listElement : list) {
+            add(listElement);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bound: " + index);
+        }
+        if (this.isEmpty()) {
+            return null;
+        }
+
+        if (index == 0) {
+            return this.head.value;
+        } else if (index == size - 1) {
+            return this.tail.value;
+        } else {
+            Node<T> currentNode = this.head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+            return currentNode.value;
+        }
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bound: " + index);
+        }
+        if (this.isEmpty()) {
+            return null;
+        }
+        Node<T> currentNode = this.head;
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bound: " + index);
+        }
+        if (this.isEmpty()) {
+            return null;
+        }
+        Node<T> currentNode = this.head;
+        if (index == 0) {
+            this.head = currentNode.next;
+            if (this.head != null) {
+                this.head.prev = null;
+            }
+        } else {
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+            if (currentNode.next == null) {
+                currentNode.prev.next = null;
+            } else {
+                currentNode.next.prev = currentNode.prev;
+            }
+            if (currentNode.prev != null) {
+                currentNode.prev.next = currentNode.next;
+            }
+        }
+        this.size--;
+        return currentNode.value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        if (this.isEmpty()) {
+            return false;
+        }
+        boolean objectIsFound = false;
+        Node<T> currentNode = this.head;
+        for (int i = 0; i < this.size; i++) {
+            if (currentNode.value == null && object == null
+                    || currentNode.value != null && currentNode.value.equals(object)) {
```

Comment:
The condition `currentNode.value == null && object == null` should be checked first to avoid a potential `NullPointerException` when calling `equals` on `currentNode.value`.